RFC/src/RFC-0121_ConsensusEncoding.md

```diff
@@ -0,0 +1,347 @@
+5. `unique_id` - [nullable] + [dynamic vector]
+6. `parent_public_key` - [nullable] + 32-byte compressed public key 
+7. `asset` - [nullable] + [AssetOutputFeatures](#assetoutputfeatures)
+8. `mint_non_fungible` - [nullable] + [MintNonFungibleFeatures](#mintnonfungiblefeatures)
+9. `sidechain_checkpoint` - [nullable] + [SideChainCheckpointFeatures](#sidechaincheckpointfeatures)
+
+##### AssetOutputFeatures
+
+- `public_key` - [RistrettoPublicKey] 
+- `template_ids` - [dynamic vector] + [varint]
+- `template_parameters` - [dynamic vector]
+
+##### MintNonFungibleFeatures
+
+- `asset_public_key` - [RistrettoPublicKey] 
+- `asset_owner_commitment` - [RistrettoPublicKey] 
+
+##### SideChainCheckpointFeatures
+
+- `merkle_root` - [fixed sized array]
+- `committee` - [dynamic vector] + [RistrettoPublicKey]
+
+#### Transaction Output
+[Transaction Output]: #transaction-output "Transaction Output"
+
+```rust,ignore
+pub struct TransactionOutput {
+    pub version: u8,
+    pub features: OutputFeatures,
+    pub commitment: Commitment,
+    pub proof: RangeProof,
+    pub script: TariScript,
+    pub sender_offset_public_key: PublicKey,
+    pub metadata_signature: ComSignature,
+    pub covenant: Covenant,
+}
+```
+
+The canonical hash commits to the common data between an output and the input spending that output
+i.e. `output_hash = Hash(version | features | commitment | script | covenant)`. 
+
+This hash is appended to the output Merkle tree.
+
+The encoding is defined as follows:
+
+- `version` - 1 byte
+- `features` - [OutputFeatures]
+- `commitment` - [RistrettoPublicKey]
+- `script` - byte length as [varint] + [TariScript]
+- `covenant` - byte length as [varint] + [Covenant]
+
+##### Witness hash
+
+The witness hash is appended to the witness Merkle tree.
+
+- `proof` - Raw proof bytes encoded using [dynamic vector] encoding
+- `metadata_signature` - [SchnorrSignature](#schnorr-signature)
+
+##### Metadata signature challenge
+
+See [Metadata Signature](./Glossary.md#metadata-signature) for details.
+
+- `public_commitment_nonce` - [RistrettoPublicKey]
+- `script` - byte length as [varint] + [TariScript]
+- `features` - [OutputFeatures]
+- `sender_offset_public_key` - [RistrettoPublicKey]
+- `commitment` - [RistrettoPublicKey]
+- `covenant`- byte length as [varint] + [Covenant]
+
+### Transaction Input
+
+The following struct represents the full transaction input data for reference. The actual input struct does not duplicate the output data
+to optimise storage and transmission of the input. 
+
+```rust,ignore
+pub struct TransactionInput {
+  pub version: u8,
+  pub input_data: ExecutionStack,
+  pub script_signature: ComSignature,
+  // Output data
+  pub output_version: TransactionOutputVersion,
+  pub features: OutputFeatures,
+  pub commitment: Commitment,
+  pub script: TariScript,
+  pub sender_offset_public_key: PublicKey,
+  pub covenant: Covenant, 
+}
+```
+
+The transaction input canonical hash pre-image is constructed as follows:
```

Comment:
I would suggest using this instead:
input_version
output_hash
input_data
script_signature

Reply:
This just reflects what is in the code, but this makes sense since the input hash can be generated even with the input in 'compact mode' - I'll update here and add a TODO to the code

Reply:
Yeah let's correct the problems in this RFC and then we'll create issues/bugs to fix the code